src/assign.c

```diff
@@ -819,89 +702,166 @@ const char *memrecycle(SEXP target, SEXP where, int start, int len, SEXP source)
+      // to assign the NA later as-is without any coerce.  Otherwise, error.
+      bool ok = true;
+      switch(TYPEOF(source)) {
+      case LGLSXP: case INTSXP: {
+        const int *sd = INTEGER(source);
+        for (int k=0; k<slen; ++k) {
+          if (sd[k]!=NA_INTEGER) { ok=false; break; }
         }
-        SET_TRUELENGTH(s, -k-1);
-      }
-      int nAdd = 0;
-      for (int k=0; k<nSourceLevels; ++k) {
-        const SEXP s = sourceLevelsD[k];
-        const int tl = TRUELENGTH(s);
-        if (tl>=0) {
-          if (tl>0) savetl(s);
-          SET_TRUELENGTH(s, -nTargetLevels-(++nAdd));
+      } break;
+      case REALSXP: {
+        const double *sd = REAL(source);
+        for (int k=0; k<slen; ++k) {
+          if (!ISNAN(sd[k])) { ok=false; break; }
         }
+      } break;
+      default:
+        ok = false;
       }
-      const int nSource = length(source);
-      const int *sourceD = INTEGER(source);
-      int *newSourceD = INTEGER(newSource);
-      for (int i=0; i<nSource; ++i) {  // convert source integers to refer to target levels
-        const int val = sourceD[i];
-        newSourceD[i] = val==NA_INTEGER ? NA_INTEGER : -TRUELENGTH(sourceLevelsD[val-1]);
-      }
-      source = newSource;
-      for (int k=0; k<nTargetLevels; ++k) SET_TRUELENGTH(targetLevelsD[k], 0);  // don't need those anymore
-      if (nAdd) {
-        // cannot grow the levels yet as that would be R call which could fail to alloc and we have no hook to clear up
-        SEXP *temp = (SEXP *)malloc(nAdd * sizeof(SEXP *));
-        if (!temp) {
-          // # nocov start
-          for (int k=0; k<nSourceLevels; ++k) SET_TRUELENGTH(sourceLevelsD[k], 0);
-          savetl_end();
-          error("Unable to allocate working memory of %d bytes to combine factor levels", nAdd*sizeof(SEXP *));
-          // # nocov end
+      if (!ok)
+        error("Cannot assign '%s' to 'factor'. Factor columns can only be assigned factor or character values, or NA in any type.", type2char(TYPEOF(source)));
+    } else {
+      // either factor or character being assigned to factor column
+      SEXP targetLevels = PROTECT(getAttrib(target, R_LevelsSymbol)); protecti++;
+      SEXP sourceLevels = source;  // character source
+      if (sourceIsFactor) { sourceLevels=PROTECT(getAttrib(source, R_LevelsSymbol)); protecti++; }
+      if (!R_compute_identical(sourceLevels, targetLevels, 0)) {
+        const int nTargetLevels=length(targetLevels), nSourceLevels=length(sourceLevels);
+        const SEXP *targetLevelsD=STRING_PTR(targetLevels), *sourceLevelsD=STRING_PTR(sourceLevels);
+        SEXP newSource = PROTECT(allocVector(INTSXP, length(source))); protecti++;
+        savetl_init();
+        for (int k=0; k<nTargetLevels; ++k) {
+          const SEXP s = targetLevelsD[k];
+          const int tl = TRUELENGTH(s);
+          if (tl>0) {
+            savetl(s);
+          } else if (tl<0) {
+            // # nocov start
```

Comment:
Yes.  Just internal errors.  If somehow factor levels contain duplicates (an error because levels must be unique in R),  the 2nd dup would already have our negative truelength assigned to it from the first one (so that's one way to get tl<0). If there are no duplicates then what I had in mind was if core R ever changes their own truelength clobber to be negative (we assume and check that R's internal usage is always positive).   This is what I was trying to get at in the error message: "internal error: levels of target are either not unique or have truelength<0".